src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombiner.java

```diff
@@ -0,0 +1,136 @@
+    var publicOE = publicDataRetrievalService.getCompanyProfile();
+
+    return Pair.of(publicOE, privateOE);
+  }
+
+  /**
+   * Creates a map of the combined Beneficial Owner data from the public and private APIs <br> Here
+   * the Hashed ID is the key, and the value is a Pair of the public data on the left and the
+   * private data on the right
+   *
+   * @return Map of the combined Beneficial Owner data
+   * @throws NoSuchAlgorithmException
+   */
+  public Map<String, Pair<PscApi, PrivateBoDataApi>> buildMergedBeneficialOwnerDataMap()
+      throws NoSuchAlgorithmException {
+
+    var combineBO = new HashMap<String, Pair<PscApi, PrivateBoDataApi>>();
+
+    var privateBOs = privateDataRetrievalService.getBeneficialOwnerData();
+    var publicPSCs = publicDataRetrievalService.getPscs();
+
+    for (var privateBO : privateBOs.getBoPrivateData()) {
+      putInPairMap(combineBO, getHashedId(privateBO.getPscId()), privateBO, Position.RIGHT);
+    }
+
+    for (var publicPSC : publicPSCs.getItems()) {
+      String[] linkAsArray = publicPSC
+          .getLinks()
+          .getSelf()
+          .split("/");
+      String hashedId = linkAsArray[linkAsArray.length - 1];
+      putInPairMap(combineBO, hashedId, publicPSC, Position.LEFT);
+    }
+
+    return combineBO;
+  }
+
+  /**
+   * Creates a map of the combined Managing Officer data from the public and private APIs <br> Here
+   * the Hashed ID is the key, and the value is a Pair of the public data on the left and the
+   * private data on the right
+   *
+   * @return Map of the combined Beneficial Owner data
+   * @throws NoSuchAlgorithmException
+   */
+  public Map<String, Pair<CompanyOfficerApi, ManagingOfficerDataApi>> buildMergedManagingOfficerDataMap()
+      throws NoSuchAlgorithmException {
+    var combineMO = new HashMap<String, Pair<CompanyOfficerApi, ManagingOfficerDataApi>>();
+
+    var privateMOs = privateDataRetrievalService.getManagingOfficerData();
+    var publicOfficers = publicDataRetrievalService.getOfficers();
+
+    for (var privateMO : privateMOs.getManagingOfficerData()) {
+      putInPairMap(combineMO, getHashedId(privateMO.getManagingOfficerId()), privateMO,
+          Position.RIGHT);
+    }
+
+    for (var publicOfficer : publicOfficers.getItems()) {
+      String[] linkAsArray = publicOfficer.getLinks().getSelf().split("/");
+      String hashedId = linkAsArray[linkAsArray.length - 1];
+      putInPairMap(combineMO, hashedId, publicOfficer, Position.LEFT);
+    }
+
+    return combineMO;
+  }
+
+  /**
+   * Puts a value into a map, either in the left or right of a pair depending on the leftRight
+   *
+   * @param map      Map to put the value into
+   * @param key      Key to put the value under
+   * @param value    Value to put into the Pair in the map
+   * @param position Whether to put the value in the left or right of the pair
+   */
+  private <L, R> void putInPairMap(Map<String, Pair<L, R>> map, String key, Object value,
```

Comment:
Can we have two methods to addPrivateData and addPublicData that don't use casts?

Reply:
I feel like that violates DRY Principles if I'm honest.
The way it's written currently there is no chance of their being an issue with casting